File: type/src/main/java/com/qmuiteam/qmui/type/TypeModel.java
/*
 * Tencent is pleased to support the open source community by making QMUI_Android available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the MIT License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.qmuiteam.qmui.type;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.qmuiteam.qmui.type.element.Element;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class TypeModel {
    private CharSequence mOrigin;
    private final Map<Integer, Element> mElementMap;
    private Element mFirstElement;
    private Element mLastElement;
    @Nullable
    private Element mFirstEffect;

    public TypeModel(
            CharSequence origin,
            @NonNull Map<Integer, Element> elementMap,
             Element firstElement,
             Element lastElement,
             @Nullable Element firstEffect) {
        mOrigin = origin;
        mElementMap = elementMap;
        mFirstElement = firstElement;
        mLastElement = lastElement;
        mFirstEffect = firstEffect;
    }

    @Nullable
    public Element getFirstEffect() {
        return mFirstEffect;
    }

    public CharSequence getOrigin() {
        return mOrigin;
    }


    public EffectRemover addBgEffect(int start, int end, final int bgColor){
        List<Integer> types = new ArrayList<>();
        types.add(TypeEnvironment.TYPE_BG_COLOR);
        return unsafeAddEffect(start, end, types, new EnvironmentUpdater() {
            @Override
            public void update(TypeEnvironment env) {
                env.setBackgroundColor(bgColor);
            }
        });
    }

    public EffectRemover addTextColorEffect(int start, int end, final int textColor){
        List<Integer> types = new ArrayList<>();
        types.add(TypeEnvironment.TYPE_TEXT_COLOR);
        return unsafeAddEffect(start, end, types, new EnvironmentUpdater() {
            @Override
            public void update(TypeEnvironment env) {
                env.setTextColor(textColor);
            }
        });
    }

    public EffectRemover addUnderLineEffect(int start, int end, final int underLineColor, final int underLineHeight){
        List<Integer> types = new ArrayList<>();
        types.add(TypeEnvironment.TYPE_BORDER_BOTTOM_WIDTH);
        types.add(TypeEnvironment.TYPE_BORDER_BOTTOM_COLOR);
        return unsafeAddEffect(start, end, types, new EnvironmentUpdater() {
            @Override
            public void update(TypeEnvironment env) {
                env.setBorderBottom(underLineHeight, underLineColor);
            }
        });
    }


    public EffectRemover unsafeAddEffect(int start, int end, List<Integer> types, EnvironmentUpdater environmentUpdater){
        Element elementStart = mElementMap.get(start);
        Element elementEnd = mElementMap.get(end);
        if(elementStart == null || elementEnd == null){
            return null;
        }
        for(Integer type: types){
            elementStart.addSaveType(type);
            elementEnd.addRestoreType(type);
        }
        elementStart.addEnvironmentUpdater(environmentUpdater);
        if(mFirstEffect == null){
            mFirstEffect = elementStart;
        }else{
            mFirstEffect.insetEffect(elementStart);
        }
        elementStart.insetEffect(elementEnd);
        return new DefaultEffectRemove(this, start, end, types, environmentUpdater);
    }

    public boolean unsafeRemoveEffect(int start, int end, List<Integer> types, EnvironmentUpdater environmentUpdater){
        Element elementStart = mElementMap.get(start);
        Element elementEnd = mElementMap.get(end);
        if(elementStart == null || elementEnd == null){
            return false;
        }
        for(Integer type: types){
            elementStart.removeSaveType(type);
            elementEnd.removeStoreType(type);
        }
        elementStart.removeEnvironmentUpdater(environmentUpdater);

        mFirstEffect = elementStart.removeFromEffectListIfNeeded(mFirstEffect);
        mFirstEffect = elementEnd.removeFromEffectListIfNeeded(mFirstEffect);
        return true;
    }

    public Element firstElement() {
        return mFirstElement;
    }

    public Element lastElement() {
        return mLastElement;
    }

    @Nullable
    public Element get(int pos){
        return mElementMap.get(pos);
    }

    public interface EffectRemover {
        void remove();
    }

    static class DefaultEffectRemove implements EffectRemover{

        private final int mStart;
        private final int mEnd;
        private final List<Integer> mTypes;
        private final EnvironmentUpdater mEnvironmentUpdater;
        private final TypeModel mTypeModel;

        public DefaultEffectRemove(
                TypeModel typeModel,
                int start,
                int end,
                List<Integer> types,
                EnvironmentUpdater environmentUpdater) {
            mTypeModel = typeModel;
            mStart = start;
            mEnd = end;
            mTypes = types;
            mEnvironmentUpdater = environmentUpdater;
        }

        @Override
        public void remove() {
            mTypeModel.unsafeRemoveEffect(mStart, mEnd, mTypes, mEnvironmentUpdater);
        }
    }
}

File: type/src/main/java/com/qmuiteam/qmui/type/TypeModel.kt
/*
 * Tencent is pleased to support the open source community by making QMUI_Android available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the MIT License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.qmuiteam.qmui.type

import android.graphics.Typeface
import com.qmuiteam.qmui.type.element.Element
import java.util.*

class TypeModel(
        val origin: CharSequence,
        private val mElementMap: Map<Int, Element>,
        private val mFirstElement: Element,
        private val mLastElement: Element) {

    var firstEffect: Element? = null

    fun addTypefaceEffect(start: Int, end: Int, typeface: Typeface): EffectRemover? {
        val types: MutableList<Int> = ArrayList()
        types.add(TypeEnvironment.TYPE_TYPEFACE)
        return unsafeAddEffect(start, end, types, object : EnvironmentUpdater {
            override fun update(env: TypeEnvironment) {
                env.typeface = typeface
            }
        })
    }

    fun addTextSizeEffect(start: Int, end: Int, textSize: Float): EffectRemover? {
        val types: MutableList<Int> = ArrayList()
        types.add(TypeEnvironment.TYPE_TEXT_SIZE)
        return unsafeAddEffect(start, end, types, object : EnvironmentUpdater {
            override fun update(env: TypeEnvironment) {
                env.textSize = textSize
            }
        })
    }

    fun addBgEffect(start: Int, end: Int, bgColor: Int): EffectRemover? {
        val types: MutableList<Int> = ArrayList()
        types.add(TypeEnvironment.TYPE_BG_COLOR)
        return unsafeAddEffect(start, end, types, object : EnvironmentUpdater {
            override fun update(env: TypeEnvironment) {
                env.backgroundColor = bgColor
            }
        })
    }

    fun addTextColorEffect(start: Int, end: Int, textColor: Int): EffectRemover? {
        val types: MutableList<Int> = ArrayList()
        types.add(TypeEnvironment.TYPE_TEXT_COLOR)
        return unsafeAddEffect(start, end, types, object : EnvironmentUpdater {
            override fun update(env: TypeEnvironment) {
                env.textColor = textColor
            }
        })
    }

    fun addUnderLineEffect(start: Int, end: Int, underLineColor: Int, underLineHeight: Int): EffectRemover? {
        val types: MutableList<Int> = ArrayList()
        types.add(TypeEnvironment.TYPE_BORDER_BOTTOM_WIDTH)
        types.add(TypeEnvironment.TYPE_BORDER_BOTTOM_COLOR)
        return unsafeAddEffect(start, end, types, object : EnvironmentUpdater {
            override fun update(env: TypeEnvironment) {
                env.setBorderBottom(underLineHeight, underLineColor)
            }
        })
    }

    fun unsafeAddEffect(start: Int, end: Int, types: List<Int>, environmentUpdater: EnvironmentUpdater): EffectRemover? {
        val elementStart = mElementMap[start]
        val elementEnd = mElementMap[end]
        if (elementStart == null || elementEnd == null) {
            return null
        }
        for (type in types) {
            elementStart.addSaveType(type)
            elementEnd.addRestoreType(type)
        }
        elementStart.addEnvironmentUpdater(environmentUpdater)
        firstEffect = if (firstEffect == null) {
            elementStart
        } else {
            elementStart.insertEffectTo(firstEffect!!)
        }
        firstEffect = elementEnd.insertEffectTo(firstEffect!!)
        return DefaultEffectRemove(this, start, end, types, environmentUpdater)
    }

    fun unsafeRemoveEffect(start: Int, end: Int, types: List<Int>, environmentUpdater: EnvironmentUpdater): Boolean {
        val elementStart = mElementMap[start]
        val elementEnd = mElementMap[end]
        if (elementStart == null || elementEnd == null) {
            return false
        }
        for (type in types) {
            elementStart.removeSaveType(type)
            elementEnd.removeStoreType(type)
        }
        elementStart.removeEnvironmentUpdater(environmentUpdater)
        firstEffect = elementStart.removeFromEffectListIfNeeded(firstEffect)
        firstEffect = elementEnd.removeFromEffectListIfNeeded(firstEffect)
        return true
    }

    fun firstElement(): Element {
        return mFirstElement
    }

    fun lastElement(): Element {
        return mLastElement
    }

    operator fun get(pos: Int): Element? {
        return mElementMap[pos]
    }

    interface EffectRemover {
        fun remove()
    }
}

class DefaultEffectRemove(
        private val typeModel: TypeModel,
        private val start: Int,
        private val end: Int,
        private val types: List<Int>,
        private val environmentUpdater: EnvironmentUpdater) : TypeModel.EffectRemover {
    override fun remove() {
        typeModel.unsafeRemoveEffect(start, end, types, environmentUpdater)
    }
}